panic: runtime error: invalid memory address or nil pointer dereference at grpc@v1.48.0/stream.go:380

NOTE: if you are reporting is a potential security vulnerability or a crash,
please follow our CVE process at
https://github.com/grpc/proposal/blob/master/P4-grpc-cve-process.md instead of
filing an issue here.

Please see the FAQ in our main README.md, then answer the questions below
before submitting your issue.

### What version of gRPC are you using?
1.48.0

### What version of Go are you using (`go version`)?
1.18.4

### What operating system (Linux, Windows, …) and version?
linux (ubuntu 18.04)

### What did you do?
If possible, provide a recipe for reproducing the error.

https://chromium.googlesource.com/infra/goma/server/+/refs/heads/main/profiler/profiler.go#34
```go
               err = profiler.Start(profiler.Config{Service: target},
                        // Disallow grpc in google-api-go-client to send stats/trace of profiler grpc's api call.
                        option.WithGRPCDialOption(grpc.WithStatsHandler(nil)))
```

### What did you expect to see?

no runtime panic (as it was in 1.47.0 or prior version)

### What did you see instead?
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x8e9e84]

goroutine 40 [running]:
google.golang.org/grpc.(*clientStream).newAttemptLocked(0xc0000ca5a0, 0x0)
	/workspace/pkg/mod/google.golang.org/grpc@v1.48.0/stream.go:380 +0x744
google.golang.org/grpc.newClientStreamWithParams({0x11c51d0, 0xc00062f650}, 0x1833f40, 0xc000359c00, {0xf2fe1a, 0x3f}, {0x0, 0x0, 0x0, 0x0, ...}, ...)
	/workspace/pkg/mod/google.golang.org/grpc@v1.48.0/stream.go:306 +0x90c
google.golang.org/grpc.newClientStream.func2({0x11c51d0?, 0xc00062f650?}, 0xc00062f650?)
	/workspace/pkg/mod/google.golang.org/grpc@v1.48.0/stream.go:192 +0x9f
google.golang.org/grpc.newClientStream({0x11c51d0, 0xc00062f650}, 0x1833f40, 0xc000359c00, {0xf2fe1a, 0x3f}, {0xc000046870, 0x1, 0x1})
	/workspace/pkg/mod/google.golang.org/grpc@v1.48.0/stream.go:220 +0x4c2
google.golang.org/grpc.invoke({0x11c51d0?, 0xc00062f650?}, {0xf2fe1a?, 0x70?}, {0xe668e0, 0xc00044e300}, {0xe99000, 0xc0006312d0}, 0x0?, {0xc000046870, ...})
	/workspace/pkg/mod/google.golang.org/grpc@v1.48.0/call.go:66 +0x7d
google.golang.org/grpc.(*ClientConn).Invoke(0x7f5fb9688108?, {0x11c51d0?, 0xc00062f650?}, {0xf2fe1a?, 0x0?}, {0xe668e0?, 0xc00044e300?}, {0xe99000?, 0xc0006312d0?}, {0xc000046870, ...})
	/workspace/pkg/mod/google.golang.org/grpc@v1.48.0/call.go:37 +0x265
google.golang.org/genproto/googleapis/devtools/cloudprofiler/v2.(*profilerServiceClient).CreateProfile(0xc000110c30, {0x11c51d0, 0xc00062f650}, 0x7f5fb9688108?, {0xc000046870, 0x1, 0x1})
	/workspace/pkg/mod/google.golang.org/genproto@v0.0.0-20220624142145-8cd45d7dbd1f/googleapis/devtools/cloudprofiler/v2/profiler.pb.go:847 +0xce
cloud.google.com/go/profiler.(*agent).createProfile.func1({0x11c51d0, 0xc00062f650}, {0xc000046860, {0x0, 0x0, 0x0}, {0x0, 0x0}})
	/workspace/pkg/mod/cloud.google.com/go/profiler@v0.3.0/profiler.go:337 +0x114
github.com/googleapis/gax-go/v2.invoke({0x11c51d0, 0xc00062f650}, 0xc00006fec8, {0xc000046860, {0x0, 0x0, 0x0}, {0x0, 0x0}}, 0xf55210)
	/workspace/pkg/mod/github.com/googleapis/gax-go/v2@v2.4.0/invoke.go:72 +0xce
github.com/googleapis/gax-go/v2.Invoke({0x11c51d0, 0xc00062f650}, 0x0?, {0xc000059eb8, 0x1, 0x0?})
	/workspace/pkg/mod/github.com/googleapis/gax-go/v2@v2.4.0/invoke.go:50 +0xa5
cloud.google.com/go/profiler.(*agent).createProfile(0xc000652400, {0x11c51d0, 0xc00062f650})
	/workspace/pkg/mod/cloud.google.com/go/profiler@v0.3.0/profiler.go:334 +0x23c
cloud.google.com/go/profiler.pollProfilerService({0x11c51d0, 0xc00062f650}, 0x1c?)
	/workspace/pkg/mod/cloud.google.com/go/profiler@v0.3.0/profiler.go:617 +0xd2
created by cloud.google.com/go/profiler.start
	/workspace/pkg/mod/cloud.google.com/go/profiler@v0.3.0/profiler.go:262 +0x6b3
```

--- a/default_dial_option_server_option_test.go
+++ b/default_dial_option_server_option_test.go
@@ -61,7 +61,7 @@ func (s) TestAddExtraDialOptions(t *testing.T) {
 func (s) TestAddExtraServerOptions(t *testing.T) {
 	const maxRecvSize = 998765
 	// Set and check the ServerOptions
-	opts := []ServerOption{StatsHandler(nil), Creds(insecure.NewCredentials()), MaxRecvMsgSize(maxRecvSize)}
+	opts := []ServerOption{Creds(insecure.NewCredentials()), MaxRecvMsgSize(maxRecvSize)}
 	internal.AddExtraServerOptions.(func(opt ...ServerOption))(opts...)
 	for i, opt := range opts {
 		if extraServerOptions[i] != opt {

--- a/dialoptions.go
+++ b/dialoptions.go
@@ -392,6 +392,7 @@ func WithDialer(f func(string, time.Duration) (net.Conn, error)) DialOption {
 func WithStatsHandler(h stats.Handler) DialOption {
 	return newFuncDialOption(func(o *dialOptions) {
 		if h == nil {
+			logger.Error("ignoring nil parameter in grpc.WithStatsHandler ClientOption")
 			// Do not allow a nil stats handler, which would otherwise cause
 			// panics.
 			return

--- a/dialoptions.go
+++ b/dialoptions.go
@@ -84,7 +84,7 @@ var extraDialOptions []DialOption
 // EmptyDialOption does not alter the dial configuration. It can be embedded in
 // another structure to build custom dial options.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This type is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -275,7 +275,7 @@ func WithBlock() DialOption {
 // the context.DeadlineExceeded error.
 // Implies WithBlock()
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -304,7 +304,7 @@ func WithInsecure() DialOption {
 // WithNoProxy returns a DialOption which disables the use of proxies for this
 // ClientConn. This is ignored if WithDialer or WithContextDialer are used.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -335,7 +335,7 @@ func WithPerRPCCredentials(creds credentials.PerRPCCredentials) DialOption {
 // the ClientConn.WithCreds. This should not be used together with
 // WithTransportCredentials.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -391,6 +391,11 @@ func WithDialer(f func(string, time.Duration) (net.Conn, error)) DialOption {
 // all the RPCs and underlying network connections in this ClientConn.
 func WithStatsHandler(h stats.Handler) DialOption {
 	return newFuncDialOption(func(o *dialOptions) {
+		if h == nil {
+			// Do not allow a nil stats handler, which would otherwise cause
+			// panics.
+			return
+		}
 		o.copts.StatsHandlers = append(o.copts.StatsHandlers, h)
 	})
 }
@@ -403,7 +408,7 @@ func WithStatsHandler(h stats.Handler) DialOption {
 // FailOnNonTempDialError only affects the initial dial, and does not do
 // anything useful unless you are also using WithBlock().
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -483,7 +488,7 @@ func WithAuthority(a string) DialOption {
 // current ClientConn's parent. This function is used in nested channel creation
 // (e.g. grpclb dial).
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -545,7 +550,7 @@ func WithMaxHeaderListSize(s uint32) DialOption {
 // WithDisableHealthCheck disables the LB channel health checking for all
 // SubConns of this ClientConn.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -592,7 +597,7 @@ func withMinConnectDeadline(f func() time.Duration) DialOption {
 // resolver.Register.  They will be matched against the scheme used for the
 // current Dial only, and will take precedence over the global registry.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.

--- a/server.go
+++ b/server.go
@@ -190,7 +190,7 @@ type ServerOption interface {
 // EmptyServerOption does not alter the server configuration. It can be embedded
 // in another structure to build custom server options.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This type is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -305,7 +305,7 @@ func CustomCodec(codec Codec) ServerOption {
 // https://github.com/grpc/grpc-go/blob/master/Documentation/encoding.md#using-a-codec.
 // Will be supported throughout 1.x.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -426,7 +426,7 @@ func ChainStreamInterceptor(interceptors ...StreamServerInterceptor) ServerOptio
 // InTapHandle returns a ServerOption that sets the tap handle for all the server
 // transport to be created. Only one can be installed.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -442,6 +442,11 @@ func InTapHandle(h tap.ServerInHandle) ServerOption {
 // StatsHandler returns a ServerOption that sets the stats handler for the server.
 func StatsHandler(h stats.Handler) ServerOption {
 	return newFuncServerOption(func(o *serverOptions) {
+		if h == nil {
+			// Do not allow a nil stats handler, which would otherwise cause
+			// panics.
+			return
+		}
 		o.statsHandlers = append(o.statsHandlers, h)
 	})
 }
@@ -469,7 +474,7 @@ func UnknownServiceHandler(streamHandler StreamHandler) ServerOption {
 // new connections.  If this is not set, the default is 120 seconds.  A zero or
 // negative value will result in an immediate timeout.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -490,7 +495,7 @@ func MaxHeaderListSize(s uint32) ServerOption {
 // HeaderTableSize returns a ServerOption that sets the size of dynamic
 // header table for stream.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -505,7 +510,7 @@ func HeaderTableSize(s uint32) ServerOption {
 // zero (default) will disable workers and spawn a new goroutine for each
 // stream.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -956,19 +961,19 @@ var _ http.Handler = (*Server)(nil)
 // To share one port (such as 443 for https) between gRPC and an
 // existing http.Handler, use a root http.Handler such as:
 //
-//   if r.ProtoMajor == 2 && strings.HasPrefix(
-//   	r.Header.Get("Content-Type"), "application/grpc") {
-//   	grpcServer.ServeHTTP(w, r)
-//   } else {
-//   	yourMux.ServeHTTP(w, r)
-//   }
+//	if r.ProtoMajor == 2 && strings.HasPrefix(
+//		r.Header.Get("Content-Type"), "application/grpc") {
+//		grpcServer.ServeHTTP(w, r)
+//	} else {
+//		yourMux.ServeHTTP(w, r)
+//	}
 //
 // Note that ServeHTTP uses Go's HTTP/2 server implementation which is totally
 // separate from grpc-go's HTTP/2 server. Performance and features may vary
 // between the two paths. ServeHTTP does not support some gRPC features
 // available through grpc-go's HTTP/2 server.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -1674,7 +1679,7 @@ type streamKey struct{}
 // NewContextWithServerTransportStream creates a new context from ctx and
 // attaches stream to it.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -1689,7 +1694,7 @@ func NewContextWithServerTransportStream(ctx context.Context, stream ServerTrans
 //
 // See also NewContextWithServerTransportStream.
 //
-// Experimental
+// # Experimental
 //
 // Notice: This type is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -1704,7 +1709,7 @@ type ServerTransportStream interface {
 // ctx. Returns nil if the given context has no stream associated with it
 // (which implies it is not an RPC invocation context).
 //
-// Experimental
+// # Experimental
 //
 // Notice: This API is EXPERIMENTAL and may be changed or removed in a
 // later release.
@@ -1825,12 +1830,12 @@ func (s *Server) getCodec(contentSubtype string) baseCodec {
 // When called multiple times, all the provided metadata will be merged.  All
 // the metadata will be sent out when one of the following happens:
 //
-// - grpc.SendHeader is called, or for streaming handlers, stream.SendHeader.
-// - The first response message is sent.  For unary handlers, this occurs when
-//   the handler returns; for streaming handlers, this can happen when stream's
-//   SendMsg method is called.
-// - An RPC status is sent out (error or success).  This occurs when the handler
-//   returns.
+//   - grpc.SendHeader is called, or for streaming handlers, stream.SendHeader.
+//   - The first response message is sent.  For unary handlers, this occurs when
+//     the handler returns; for streaming handlers, this can happen when stream's
+//     SendMsg method is called.
+//   - An RPC status is sent out (error or success).  This occurs when the handler
+//     returns.
 //
 // SetHeader will fail if called after any of the events above.
 //

--- a/server.go
+++ b/server.go
@@ -443,6 +443,7 @@ func InTapHandle(h tap.ServerInHandle) ServerOption {
 func StatsHandler(h stats.Handler) ServerOption {
 	return newFuncServerOption(func(o *serverOptions) {
 		if h == nil {
+			logger.Error("ignoring nil parameter in grpc.StatsHandler ServerOption")
 			// Do not allow a nil stats handler, which would otherwise cause
 			// panics.
 			return

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -8045,6 +8045,7 @@ func (s) TestServerClosesConn(t *testing.T) {
 // TestNilStatsHandler ensures we do not panic as a result of a nil stats
 // handler.
 func (s) TestNilStatsHandler(t *testing.T) {
+	grpctest.TLogger.ExpectErrorN("ignoring nil parameter", 2)
 	ss := &stubserver.StubServer{
 		UnaryCallF: func(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
 			return &testpb.SimpleResponse{}, nil

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -8050,7 +8050,7 @@ func (s) TestNilStatsHandler(t *testing.T) {
 			return &testpb.SimpleResponse{}, nil
 		},
 	}
-	if err := ss.Start([]grpc.ServerOption{}, grpc.WithStatsHandler(nil)); err != nil {
+	if err := ss.Start([]grpc.ServerOption{grpc.StatsHandler(nil)}, grpc.WithStatsHandler(nil)); err != nil {
 		t.Fatalf("Error starting endpoint server: %v", err)
 	}
 	defer ss.Stop()

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -8042,6 +8042,26 @@ func (s) TestServerClosesConn(t *testing.T) {
 	t.Fatalf("timed out waiting for conns to be closed by server; still open: %v", atomic.LoadInt32(&wrapLis.connsOpen))
 }
 
+// TestNilStatsHandler ensures we do not panic as a result of a nil stats
+// handler.
+func (s) TestNilStatsHandler(t *testing.T) {
+	ss := &stubserver.StubServer{
+		UnaryCallF: func(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+			return &testpb.SimpleResponse{}, nil
+		},
+	}
+	if err := ss.Start([]grpc.ServerOption{}, grpc.WithStatsHandler(nil)); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if _, err := ss.Client.UnaryCall(ctx, &testpb.SimpleRequest{}); err != nil {
+		t.Fatalf("Unexpected error from UnaryCall: %v", err)
+	}
+}
+
 // TestUnexpectedEOF tests a scenario where a client invokes two unary RPC
 // calls. The first call receives a payload which exceeds max grpc receive
 // message length, and the second gets a large response. This second RPC should

